platforms/core-runtime/functional/src/main/java/org/gradle/internal/serialization/Cached.java

```diff
@@ -53,16 +60,17 @@ public T get() {
 
         private Try<T> result() {
             if (result == null) {
-                result = tryComputation();
+                // copy reference into the call stack to avoid exacerbating https://github.com/gradle/gradle/issues/31239
+                result = tryComputation(computation);
                 computation = null;
             }
             return result;
         }
 
         @Nonnull
-        private Try<T> tryComputation() {
+        private Try<T> tryComputation(Callable<T> toCompute) {
```

Comment:
Without this change, the added test case would fail very frequently. Now, it fails just as rarely as before we started wrapping the `computation` reference into a lambda, which seems to be the trigger.

Reply:
Since the intent is not to fix it yet but to make it a little safer, I'd capture the `computation` field at the beginning of the method, before entering `if (result == null)` body. The method could even throw if `computation == null` inside the body.

Reply:
To be clear, without the fix, the test fails 2-2.5% of the time (10000 runs leads do 200+ failures). 

I could reproduce the failure by running 100k iterations with the fix and your proposed change. With your proposed changes, plus adding `volatile` to `result` and `computation`, I could not reproduce the NPE even with 1M iterations.

Reply:
Any concerns with making both variable `volatile`?

Reply:
To be honest, I didn't get to see it running to 1M times, as Gradle slows down quite a bit after 400K.